lib/micromachine.rb

```diff
@@ -44,9 +49,21 @@ def states
 private
 
   def change(event)
-    @state = transitions_for[event][@state]
+    previous_state, @state = @state, transitions_for[event][@state]
+
+    all_arguments = [event, previous_state, state]
+
     callbacks = @callbacks[@state] + @callbacks[:any]
-    callbacks.each { |callback| callback.call(event) }
+    callbacks.each do |callback|
+      case callback.arity
+      when -1
+        callback.call(*all_arguments)
+      when 0..3
+        arguments = all_arguments.take(callback.arity)
```

Comment:
Yep. 

To reduce the number of array allocations during the `each` invocations. But it just came to me, that `take` will allocate new array for result, so yeah, this is no use. 

Alternative is to  expand the `case` to explicitly match each number of arguments, in this case we can avoid array allocation, but at the cost of the additional branching.